File: src/main/java/me/ste/ivremotecontrol/block/signalcontrollerinterface/SignalControllerInterfaceTileEntity.java
package me.ste.ivremotecontrol.block.signalcontrollerinterface;

import li.cil.oc.api.Network;
import li.cil.oc.api.machine.Arguments;
import li.cil.oc.api.machine.Callback;
import li.cil.oc.api.machine.Context;
import li.cil.oc.api.network.Connector;
import li.cil.oc.api.network.Visibility;
import li.cil.oc.api.prefab.TileEntityEnvironment;
import mcinterface1122.BuilderTileEntity;
import minecrafttransportsimulator.blocks.tileentities.components.ATileEntityBase;
import minecrafttransportsimulator.blocks.tileentities.instances.TileEntitySignalController;
import minecrafttransportsimulator.mcinterface.MasterLoader;
import minecrafttransportsimulator.packets.instances.PacketTileEntitySignalControllerChange;
import net.minecraft.block.BlockDirectional;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ITickable;
import net.minecraft.util.math.BlockPos;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

public class SignalControllerInterfaceTileEntity extends TileEntityEnvironment implements ITickable {
    private boolean hasEnergy = false;

    public SignalControllerInterfaceTileEntity() {
        this.node = Network.newNode(this, Visibility.Network).withConnector().withComponent("signalcontroller").create();
    }

    @Override
    public void update() {
        if(this.node != null) {
            this.hasEnergy = ((Connector) this.node).tryChangeBuffer(-0.5D);
        }
    }

    public TileEntitySignalController getController() {
        if(!this.hasEnergy) {
            return null;
        }
        BlockPos loaderPos = this.pos.offset(this.world.getBlockState(this.pos).getValue(BlockDirectional.FACING));
        TileEntity tileEntity = this.world.getTileEntity(loaderPos);
        if(!(tileEntity instanceof BuilderTileEntity)) {
            return null;
        }
        ATileEntityBase<?> tileEntityBase;
        try {
            Field field = BuilderTileEntity.class.getDeclaredField("tileEntity");
            field.setAccessible(true);
            tileEntityBase = (ATileEntityBase<?>) field.get(tileEntity);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            return null;
        }
        if(!(tileEntityBase instanceof TileEntitySignalController)) {
            return null;
        }
        return (TileEntitySignalController) tileEntityBase;
    }

    @Callback(direct = true, doc = "function(): boolean -- check whether the signal controller is available")
    public Object[] isAvailable(Context ctx, Arguments args) {
        return new Object[] {this.getController() != null};
    }

    @Callback(direct = true, doc = "function(): string -- get the controller primary axis. Can be either Z or X")
    public Object[] getAxis(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.mainDirectionXAxis ? "X" : "Z"};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(axis:string) -- set the controller primary axis. Can be either Z or X")
    public Object[] setAxis(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            String axis = args.checkString(0);
            boolean modified = false;
            if(axis.equalsIgnoreCase("Z")) {
                controller.mainDirectionXAxis = false;
                modified = true;
            } else if(axis.equalsIgnoreCase("X")) {
                controller.mainDirectionXAxis = true;
                modified = true;
            }
            if(modified) {
                controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
                MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
            }
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): string -- get the current controller mode. Can be either TIMED_CYCLE, VEHICLE_TRIGGER, REDSTONE_TRIGGER or REMOTE_CONTROL")
    public Object[] getMode(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.currentOpMode.name()};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(mode:string) -- set the controller mode. Can be either TIMED_CYCLE, VEHICLE_TRIGGER, REDSTONE_TRIGGER or REMOTE_CONTROL")
    public Object[] setMode(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            try {
                controller.currentOpMode = TileEntitySignalController.OpMode.valueOf(args.checkString(0));
                controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
                MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
            } catch(IllegalArgumentException ignored) {}
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): number -- get the time main signal is green")
    public Object[] getGreenMainTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.greenMainTime};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(time:number) -- set the time main signal is green")
    public Object[] setGreenMainTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            controller.greenMainTime = args.checkInteger(0);
            controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
            MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): number -- get the time cross signal is green")
    public Object[] getGreenCrossTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.greenCrossTime};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(time:number) -- set the time cross signal is green")
    public Object[] setGreenCrossTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            controller.greenCrossTime = args.checkInteger(0);
            controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
            MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): number -- get the time main signal is yellow")
    public Object[] getYellowMainTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.yellowMainTime};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(time: number) -- set the time main signal is yellow")
    public Object[] setYellowMainTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            controller.yellowMainTime = args.checkInteger(0);
            controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
            MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): number -- get the time cross signal is yellow")
    public Object[] getYellowCrossTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.yellowCrossTime};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(time: number) -- set the time cross signal is yellow")
    public Object[] setYellowCrossTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            controller.yellowCrossTime = args.checkInteger(0);
            controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
            MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): number -- get the time all signals are red")
    public Object[] getAllRedTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.allRedTime};
        }
        return new Object[] {};
    }

    @Callback(doc = "function(time: number) -- set the time all signals are red")
    public Object[] setAllRedTime(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            controller.allRedTime = args.checkInteger(0);
            controller.updateState(TileEntitySignalController.OpState.GREEN_MAIN_RED_CROSS, true);
            MasterLoader.networkInterface.sendToAllClients(new PacketTileEntitySignalControllerChange(controller));
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): boolean -- get whether the street lights are on")
    public Object[] getLights(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.lightsOn};
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): table -- get the current signal controller state")
    public Object[] getState(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            Map<String, Object> state = new HashMap<>();
            state.put("name", controller.currentOpState.name());
            state.put("mainSignal", controller.currentOpState.mainSignalState.name());
            state.put("crossSignal", controller.currentOpState.crossSignalState.name());
            return new Object[] {state};
        }
        return new Object[] {};
    }

    @Callback(direct = true, doc = "function(): number -- get the time since the last operation has started")
    public Object[] getTimeOperationStarted(Context ctx, Arguments args) {
        TileEntitySignalController controller = this.getController();
        if(controller != null) {
            return new Object[] {controller.timeOperationStarted};
        }
        return new Object[] {};
    }
}

File: src/main/kotlin/me/ste/ivremotecontrol/block/signalcontrollerinterface/SignalControllerInterfaceTileEntity.kt
package me.ste.ivremotecontrol.block.signalcontrollerinterface

import dan200.computercraft.api.lua.ArgumentHelper
import dan200.computercraft.api.lua.ILuaContext
import dan200.computercraft.api.peripheral.IComputerAccess
import mcinterface1122.BuilderTileEntity
import me.ste.ivremotecontrol.block.peripheral.PeripheralTileEntity
import me.ste.ivremotecontrol.util.mtsTileEntity
import minecrafttransportsimulator.blocks.components.ABlockBase
import minecrafttransportsimulator.blocks.tileentities.instances.TileEntitySignalController
import minecrafttransportsimulator.mcinterface.InterfaceManager
import minecrafttransportsimulator.packets.instances.PacketTileEntitySignalControllerChange
import net.minecraft.block.BlockDirectional

class SignalControllerInterfaceTileEntity : PeripheralTileEntity("signalcontroller") {
    private val controller: TileEntitySignalController?
        get() {
            val decorPos = this.pos.offset(world.getBlockState(this.pos).getValue(BlockDirectional.FACING))
            val tileEntity = world.getTileEntity(decorPos)
            if (tileEntity !is BuilderTileEntity<*>) {
                return null
            }
            return tileEntity.mtsTileEntity as? TileEntitySignalController
        }

    private fun isAvailable(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any> =
        arrayOf(this.controller != null)

    private fun getMainDirection(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(it.mainDirectionAxis.name.toLowerCase()) }

    private fun setMainDirection(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            val direction = ArgumentHelper.getString(args, 0)

            when (direction.toLowerCase()) {
                "north" -> it.mainDirectionAxis = ABlockBase.Axis.NORTH
                "east" -> it.mainDirectionAxis = ABlockBase.Axis.EAST
                "northeast" -> it.mainDirectionAxis = ABlockBase.Axis.NORTHEAST
                "northwest" -> it.mainDirectionAxis = ABlockBase.Axis.NORTHWEST
                else -> throw ArgumentHelper.badArgument(0, "a main direction", direction)
            }

            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))

            null
        }

    private fun getMode(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(if(it.timedMode) "timed_cycle" else "vehicle_trigger") }

    private fun setMode(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            val modeString = ArgumentHelper.getString(args, 0)

            when (modeString) {
                "timed_cycle" -> it.timedMode = true
                "vehicle_trigger" -> it.timedMode = false
                else -> throw ArgumentHelper.badArgument(0, "an opmode", modeString)
            }

            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))

            null
        }

    private fun getGreenMainTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(it.greenMainTime) }

    private fun getGreenCrossTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(it.greenMainTime) }

    private fun getYellowMainTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(it.yellowMainTime) }

    private fun getYellowCrossTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(it.yellowCrossTime) }

    private fun getAllRedTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(it.allRedTime) }

    private fun setGreenMainTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            it.greenMainTime = ArgumentHelper.getInt(args, 0)
            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))
            null
        }

    private fun setGreenCrossTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            it.greenCrossTime = ArgumentHelper.getInt(args, 0)
            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))
            null
        }

    private fun setYellowMainTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            it.yellowMainTime = ArgumentHelper.getInt(args, 0)
            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))
            null
        }

    private fun setYellowCrossTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            it.yellowCrossTime = ArgumentHelper.getInt(args, 0)
            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))
            null
        }

    private fun setAllRedTime(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            it.allRedTime = ArgumentHelper.getInt(args, 0)
            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))
            null
        }

    @Deprecated("Removed")
    private fun getState(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(
            hashMapOf(
                "name" to "green_main_red_cross",
                "mainSignal" to "golight",
                "crossSignal" to "stoplight"
            )
        ) }

    @Deprecated("Removed")
    private fun getLights(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(false) }

    @Deprecated("Removed")
    private fun getTimeOperationStarted(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let { arrayOf(0) }

    @Deprecated("Use getMainDirection instead")
    private fun getAxis(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            arrayOf(
                when (it.mainDirectionAxis) {
                    ABlockBase.Axis.NORTH -> "z"
                    ABlockBase.Axis.EAST -> "x"
                    ABlockBase.Axis.NORTHEAST -> "x"
                    ABlockBase.Axis.NORTHWEST -> "z"
                    else -> "x"
                }
            )
        }

    @Deprecated("Use setMainDirection instead")
    private fun setAxis(pc: IComputerAccess, ctx: ILuaContext, args: Array<Any>): Array<Any>? =
        this.controller?.let {
            val axis = ArgumentHelper.getString(args, 0)

            when (axis.toLowerCase()) {
                "x" -> it.mainDirectionAxis = ABlockBase.Axis.EAST
                "z" -> it.mainDirectionAxis = ABlockBase.Axis.NORTH
                else -> throw ArgumentHelper.badArgument(0, "an axis", axis)
            }

            it.initializeController(null)
            InterfaceManager.packetInterface.sendToAllClients(PacketTileEntitySignalControllerChange(it))

            null
        }

    init {
        this.methods["isAvailable"] = this::isAvailable
        this.methods["getMainDirection"] = this::getMainDirection
        this.methods["setMainDirection"] = this::setMainDirection
        this.methods["getMode"] = this::getMode
        this.methods["setMode"] = this::setMode
        this.methods["getGreenMainTime"] = this::getGreenMainTime
        this.methods["getGreenCrossTime"] = this::getGreenCrossTime
        this.methods["getYellowMainTime"] = this::getYellowMainTime
        this.methods["getYellowCrossTime"] = this::getYellowCrossTime
        this.methods["getAllRedTime"] = this::getAllRedTime
        this.methods["setGreenMainTime"] = this::setGreenMainTime
        this.methods["setGreenCrossTime"] = this::setGreenCrossTime
        this.methods["setYellowMainTime"] = this::setYellowMainTime
        this.methods["setYellowCrossTime"] = this::setYellowCrossTime
        this.methods["setAllRedTime"] = this::setAllRedTime

        this.methods["getState"] = this::getState
        this.methods["getLights"] = this::getLights
        this.methods["getTimeOperationStarted"] = this::getTimeOperationStarted
        this.methods["getAxis"] = this::getAxis
        this.methods["setAxis"] = this::setAxis
    }
}